### ISSUE TITLE
Add support for joysticks and gamepads as a trigger mechanism

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#412]
https://github.com/sharpdx/SharpDX
https://github.com/sharpdx/SharpDX-Samples/blob/master/Desktop/DirectInput/JoystickApp/Program.cs

This method of monitoring for joystick/gamepad button presses receives input even when the app is in the background.

**Comment by JuliusSweetland** [comment]
https://github.com/sharpdx/SharpDX

**Comment by annakirkpatrick** [comment]
I second this feature request. I would immediately switch to using joystick/gamepad as a trigger mechanism if it was implemented.

Some information from a user perspective (perhaps relevant for motivation and/or future user documentation): 
Three common USB switch interfaces which support stimulating joystick/gamepad import are [Swifty by Origin Instruments](http://www.orin.com/access/swifty/) and the [X-keys 3](http://xkeys.com/xkeys/xk3swi.php) and [12 switch interfaces](http://xkeys.com/xkeys/xkswi.php). These are the first three (distinct) results if you do a Google search for "USB switch interface," so I think it is fair to say that many, if not most, switch input users will have a device that supports this type of input.

I personally own the X keys devices and would be happy to do beta testing.

**Comment by annakirkpatrick** [comment]
I want to document another reason why this functionality would be very useful.  In addition to the fact that Windows key combinations do not work with the keyboard key trigger mechanism (and using a mouseclick as trigger mechanism has the potential to unintentionally change the focus and cause all sorts of confusion), there is the issue of the target application trying to interpret the keypress that you are using as a trigger mechanism for OptiKey.

Let me give a concrete example. I recently installed git on my Windows 10 machine. It comes bundled with git bash. Since the Windows command prompt is pretty abysmal, and since I work in a Linux terminal all the time (yes, with OptiKey. ssh is amazing!), I tried to use git bash. Well, the default terminal emulator for git bash is mintty, and I could not find a single keystroke that mintty would ignore. I usually use F13 as my trigger key. Since F13 doesn't appear on standard keyboards, most applications will ignore it, but not mintty. It interpreted that keystroke as '~'. I spent over an hour looking for a key that would be ignored by mintty (and ideally by all the other applications I use), but I could not find one.

I have now switched to using Windows Power Shell, and that is working, so I don't need terminal emulator suggestions or anything like that.

But it would be nice to not have to go through this dance of "what keystroke will this application ignore?" every few months when I try use a new application that happens to interpret whatever key I'm using as a trigger key.

I know that switching to joystick/gamepad input won't eliminate this problem entirely, but it should be a big improvement over keyboard and mouse input.

**Comment by JuliusSweetland** [comment]
Thanks Anna. That's really useful. This feature is on my todo list at some
point.

**Comment by annakirkpatrick** [comment]
Dropping back in to this thread to document another reason why joystick/gamepad as trigger mechanism would be useful.

My research group uses Slack for communication. Though I often fall back to keyboard-only navigation, Slack's keyboard navigation is downright terrible, and so I've been using Optikey's mouse emulation tools.  

It is impossible for me to highlight and copy text. I can highlight text, using a double-click or click and drag. But, when I try to right-click to bring up the context menu to copy, my highlight disappears. In fact, I have figured out that Slack interprets almost any keystroke (exceptions are navigation keystrokes like tab and space, which obviously don't make good trigger keys) as a reason to move focus to the message composition box (removing all highlighting in the process).

Needless to say, I'm very frustrated. Admittedly, this is mostly on Slack for poor and inaccessible design. But a workaround would sure be great!

(For whatever it's worth, I'm using F 23 as a trigger key.)

**Comment by amywhays** [comment]
I'm just started using Gopher360 with an Xbox one controller on a Windows 10 Pro PC.   It seems to work well with OptiKey so far.    I set OptiKey to accept the mouse position and mouse button for the Pointer Device Source and Key Selection Source respectively in the Management Console.  I don't have an eye tracker yet. 

Gopher360: https://github.com/Tylemagne/Gopher360

**Comment by kmcnaught** [comment]
Just to say I'm working on a solution for this in case anyone else was interested. So far I have a trigger source working for XInput devices (Xbox + other recent gamepads), and I'm working on DirectInput (older gamepad API) now. If anyone wants something for early testing let me know otherwise will raise a PR sometime in January hopefully.

**Comment by kmcnaught** [comment]
I'm trying to figure out how best to manage settings for different controllers. I mainly want to document my thought processes with the github rubber duck, but would also appreciate input from @JuliusSweetland and @annakirkpatrick. 

## TL;DR 
How should we specify which controller is to be used?

## Background and thought process
We have two different APIs to support - XInput is the `"the new input standard for both the Xbox and Windows"` and DirectInput is the "legacy" API which many older / non-xbox devices use. Both can be accessed via the [SharpDX](http://sharpdx.org/) library.

In my first prototype you could only connect to "the first XInput controller" or "the last found DirectInput controller" but I am assuming that it is a requirement that users can choose the specific controller they want to use. This would be necessary if for instance someone has a joystick for cursor control + a separate switch interface/gamepad for switch control / as a trigger. Or someone may have a virtual controller installed by other software which presents itself in the list of available controllers, but they wouldn't want that shadowing the real controller they've just plugged in. 

The way you specify a controller to connect to differs between the two APIs. XInput supports up to 4 controllers connected at any time, selected via "user index" one through four, which corresponds to USB input port. There isn't an official way to get a repeatable ID from a particular controller, though it may be possible to hook into some undocumented functions as per here: https://stackoverflow.com/questions/64251320/is-it-possible-to-get-an-xinput-devices-name-product-id-vendor-id-or-some-oth

DirectInput supports a variety of devices. I am only considering those registered as "joystick" or "gamepad" but if anyone wanted to use another input device (flight stick? driving controller?) it wouldn't be hard to add in the future - I just don't have any for testing. Each device presents itself with a `ProductGuid` (which will be the same for two identical controllers) and an `InstanceGuid` which is unique per [controller+PC] combo, regardless of which port it is plugged into. Apparently it may change depending on Windows user account but I don't think that matters here. 

So it seems that for any controller trigger source we would need to specify:
- Which API to use (DirectInput / XInput)
- Which controller to use (DirectInput: UserIndex 1 to 4; XInput: unique GUID chosen via friendly description)
- Which button/key to use - for XInput these are an enum (like A,B,X,Y,LeftTrigger) and for DirectInput just a button index from 1-256. 

Currently the controller-selection looks like this: 

### XInput
(including "connected" label to help choose)
![image](https://user-images.githubusercontent.com/12151404/149670617-32853359-bc22-4d56-9c20-b383cd40f7bc.png)
### DirectInput
(including short SHA to help distinguish two identical controllers (just in case..)
![image](https://user-images.githubusercontent.com/12151404/149675021-677a52cc-417c-419b-87be-29a94c652c87.png)

I think this is reasonable so far, but please let me know if you have better suggestions. 

### Question 
The remaining question is whether (+ how) we should support a simpler default option for "I will only plug in one controller at a time - use this" or maybe "use the superset of all controllers". If a user swaps out one controller for another, would they expect it to work still? Or if they plugged the same controller into a different USB port, it would be nice if Optikey continues to pick up the same input. Or am I over-engineering things?

**Comment by annakirkpatrick** [comment]
Maybe over-engineering a little.  :-)

Of all the concerns you raise, the only one that really strikes me as an issue is plugging the same controller in to a different USB port. I can see that I might have done this when I used to work on a laptop while I was a student.  (I never got OptiKey working well on the laptop screen, but I did use the laptop for reading things when away from home and then plug into an external monitor when I got home. When plugging everything back in, my switch interface easily could have been on a different port.)  It's not a big deal to have to set it up again, but it could be annoying.  In fact, Dragon NaturallySpeaking had this exact issue in the past.  I was always having to set up my microphone again because I used a different USB port/hub combination.  It got old really fast.

I think "I will only plug in one controller at a time - use this" could be more confusing than helpful.  Since various software programs can create virtual controllers (including some other AT programs), I can see a user saying "but I only have 1 controller plugged in" when really there's a virtual software controller, too.  Also some AT hardware may act like a controller without looking like a controller.

I think there might be some value in "use the superset of all controllers."  The potential usefulness here is that if a controller is plugged into a different USB port or a different physical controller is plugged in (say of the same kind), when everything would still work without changing any settings.  And this would still work if multiple happened to be plugged in or if there was an inactive virtual controller that the user didn't know about.  The behavior also mirrors the mouse and keyboard trigger options.  (If you have multiple keyboards plugged in, you can press the trigger key on any of them.) So, I'd say that's probably a good default....but I could be convinced that I'm wrong.

**Comment by kmcnaught** [comment]
The USB port thing was bothering me, too so I've been doing some more testing. Thankfully I think my assertion about USB ports was wrong. The docs say:

> The XInput API supports up to four controllers connected at any time. The XInput functions all require a dwUserIndex parameter that is passed in to identify the controller being set or queried. This ID will be in the range of 0-3 and is set automatically by XInput. The number corresponds to the port that the controller is plugged into, and is not modifiable.

However, I've just been testing with various combinations of two controllers in two different ports, either plugged in or not. It seems that in the case where a single controller is plugged in, it always gets given the first index (actually `UserIndex.One` in SharpDX - not zero-indexed). If both are plugged in and turned on before being queried with SharpDX then their ordering seems consistent according to USB port, but if you unplug `UserIndex.One` then the remaining controller keeps at `UserIndex.Two`. I think there's a little more nuance to it but it's promising that in the situation where someone only ever has one controller they should be able to swap ports happily. And it probably would most of the time work okay with any virtual controllers, since the ordering should be consistent (say for example your virtual controller takes slot one, you see in settings that slot one and two are connected, choose slot two and probably this will always be your physical plugged-in controller). 

I think you're probably right on the superset default. I'll have a think about how that will work architecturally. Interestingly XInput offers the option `UserIndex.Any` but it doesn't appear to actually work in SharpDX.

**Comment by kmcnaught** [comment]
Available in 3.2.2 🎉
